In [59]:
import pandas as pd
# pd.set_option('max columns', 20)

# 1. IT Salary Survey EU 2018
Read the CSV file using pandas

In [60]:
df_2018 = pd.read_csv("IT Salary Survey EU 2018.csv")

## 1. Understanding the data
* **Shape** 
    * ```df.shape``` - tuple that tells us the size of our data frame. Row * Column (Like the order of a matrix)
* **Row Analysis** 
    * ```df.head()``` - function that shows the first 5 rows
    * ```df.tail()``` - function that shows the last 5 rows
* **Column Analysis** 
    * ```df.columns``` - used to list all the columns
    * ```df.dtypes``` - determines the data types of all the columns
    * ```df.describe``` - my favorite! Gives small numeric data or statistics about the columns

### 1. Shape Analysis
Find out the shape of the data frame using ```df.shape```. Note that this is not a function but a pre-stored tuple.

In [61]:
print(df_2018.shape)

(765, 14)


### 2. Row analysis

Print the first and last 5 lines of the csv data to know the size of the data we're working with

+ First five:

In [62]:
# First 5
print(df_2018.head())

             Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68

+ Last 5:

In [63]:
# Last 5
print(df_2018.tail())

               Timestamp   Age Gender     City               Position  \
760  03/06/2020 20:12:51  40.0      M     Köln  Java Developer junior   
761  28/07/2020 04:03:13   NaN      M     Köln      E.g. C# Developer   
762  28/07/2020 04:03:20   NaN      M     Köln      E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                    NaN   
764  21/09/2020 01:47:10  31.0      F  München                    Pm    

     Years of experience Your level  Current Salary  Salary one year ago  \
760                  1.0     Junior         44000.0              40000.0   
761                  1.0     Junior         45000.0              40000.0   
762                  1.0     Junior         45000.0              40000.0   
763                  NaN        NaN             NaN                  NaN   
764                 10.0     Senior        110000.0                  NaN   

     Salary two years ago Are you getting any Stock Options?  \
760               40000.0               

Since our first entry is at index 0, we have 764 entries in total. This concurs with the df.shape we ran

### 2. Column analysis
+ As we can see from the shape, we have 14 columns. What are they?

In [64]:
print(df_2018.columns)

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')


+ Determine the data type of the columns

In [65]:
print(df_2018.dtypes)

Timestamp                              object
Age                                   float64
Gender                                 object
City                                   object
Position                               object
Years of experience                   float64
Your level                             object
Current Salary                        float64
Salary one year ago                   float64
Salary two years ago                  float64
Are you getting any Stock Options?     object
Main language at work                  object
Company size                           object
Company type                           object
dtype: object


+ Get insights on the numeric data/statistics about the columns

In [66]:
df_2018.describe()

Age  Years of experience  Current Salary  Salary one year ago  \
count  672.000000           732.000000      750.000000           596.000000   
mean    32.183036             8.548497    68381.765333         62187.278523   
std      5.107268             4.729557    21196.306557         20163.008663   
min     21.000000             0.000000    10300.000000         10001.000000   
25%     29.000000             5.000000    57000.000000         52000.000000   
50%     32.000000             8.000000    65000.000000         60000.000000   
75%     35.000000            11.000000    75000.000000         70000.000000   
max     60.000000            38.000000   200000.000000        200000.000000   

       Salary two years ago  
count            463.000000  
mean           58013.475162  
std            20413.048908  
min            10001.000000  
25%            48000.000000  
50%            56000.000000  
75%            67000.000000  
max           150000.000000

## 2. Preparing the Data (Data Cleaning)
This steps involves:
+ Dropping irrelevant columns and rows
+ Identifying duplicated columns
+ Renaming columns
+ Feature Creation

Running ```df.head()``` just to remember our columns and how our data set looks like

In [67]:
df_2018.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68000.0                                 No   
1               55000.0                                 No   
2               54000.0                                 No   
3               45000.0                                Yes   
4               52000.0                                 No   

  Main language at work Company size   Company type  
0               Deutsch     100-1000        Product  
1               Deutsch       50-100        Product  
2               Deutsch        1000+        Product  
3               English        1000+        Product  
4               English     100-1000  Ecom retailer

### i. Dropping irrelevant rows or columns
There are two ways of doing this
+ Using the drop function
+ List subsetting (preferred)

#### The drop function
Say that we find the "Are you getting any stock options?" data irrelevant. We can remove it using the drop function ```df.columns.drop()```

In [68]:
df_2018.columns.drop("Are you getting any Stock Options?")

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Main language at work', 'Company size',
       'Company type'],
      dtype='object')

If you want to drop more than one column, you can do so by using a list.

In [69]:
df_2018.columns.drop(["Salary one year ago", "Salary two years ago"])

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')

Note how "Are you getting any stock options?" is back yet we had dropped it. If you wanted to analyze data without that column, you should do it in the same cell you dropped it.

#### List Subsetting
Steps:
+ Run ```df.columns``` to see all the columns
+ Copy and paste the list of columns into a cell
+ Comment out the columns you don't want and run ```df[[list of columns]]``` (The two square brackets are **extremely** important)

In [70]:
df_2018.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')

In [71]:
df_2018[['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', #'Salary one year ago',
       #'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type']]

Timestamp   Age Gender     City                       Position  \
0    14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1    14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2    14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3    14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4    14/12/2018 12:50:31  39.0      M  München                    UX Designer   
..                   ...   ...    ...      ...                            ...   
760  03/06/2020 20:12:51  40.0      M     Köln          Java Developer junior   
761  28/07/2020 04:03:13   NaN      M     Köln              E.g. C# Developer   
762  28/07/2020 04:03:20   NaN      M     Köln              E.g. C# Developer   
763  26/08/2020 09:06:44   NaN    NaN      NaN                            NaN   
764  21/09/2020 01:47:10  31.0      F  München                            Pm    

     Years of experience Your level  Current Salary Main language at work  \
0                   11.0     Senior         77000.0               Deutsch   
1                    8.0     Senior         65000.0               Deutsch   
2                   10.0     Senior         88000.0               Deutsch   
3                    6.0     Senior         78000.0               English   
4                   10.0     Senior         69000.0               English   
..                   ...        ...             ...                   ...   
760                  1.0     Junior         44000.0               Deutsch   
761                  1.0     Junior         45000.0               Deutsch   
762                  1.0     Junior         45000.0               Deutsch   
763                  NaN        NaN             NaN                   NaN   
764                 10.0     Senior        110000.0               English   

    Company size   Company type  
0       100-1000        Product  
1         50-100        Product  
2          1000+        Product  
3          1000+        Product  
4       100-1000  Ecom retailer  
..           ...            ...  
760        10-50        Product  
761        10-50        Product  
762        10-50        Product  
763          NaN            NaN  
764        1000+        Product  

[765 rows x 11 columns]

However, just like using the drop function, this only applies to the current cell and doesn't transcend to later cells. This can be proven by running ```df.head()``` in a new cell.

In [72]:
df_2018.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68000.0                                 No   
1               55000.0                                 No   
2               54000.0                                 No   
3               45000.0                                Yes   
4               52000.0                                 No   

  Main language at work Company size   Company type  
0               Deutsch     100-1000        Product  
1               Deutsch       50-100        Product  
2               Deutsch        1000+        Product  
3               English        1000+        Product  
4               English     100-1000  Ecom retailer

However, to keep working with the new subsetted data in new cells, you'll need to reassign it to the ```df``` variable or assign it to a new variable eg. ```df_subset_1```

In [73]:
df_subset_1 = df_2018[
       ['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', #'Salary one year ago',
       #'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type']
       ]

Now we can work with this subsetted data in a new cell. To prove this, we can run ```df.shape``` or ```df.head()``` to observe the changes in the columns that were applied in the previous cell.

In [74]:
df_subset_1.shape

(765, 11)

We had 14 columns but now we have 11.

In [75]:
df_subset_1.head()

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary Main language at work  \
0                 11.0     Senior         77000.0               Deutsch   
1                  8.0     Senior         65000.0               Deutsch   
2                 10.0     Senior         88000.0               Deutsch   
3                  6.0     Senior         78000.0               English   
4                 10.0     Senior         69000.0               English   

  Company size   Company type  
0     100-1000        Product  
1       50-100        Product  
2        1000+        Product  
3        1000+        Product  
4     100-1000  Ecom retailer

In [76]:
df_subset_1.dtypes

Timestamp                 object
Age                      float64
Gender                    object
City                      object
Position                  object
Years of experience      float64
Your level                object
Current Salary           float64
Main language at work     object
Company size              object
Company type              object
dtype: object

Data Preparation also involves **changing the dtype** of necessary columns. As we can see from above, timestamp is an object yet it should be a datetime. This can be changed

In [77]:
pd.to_datetime(df_2018["Timestamp"])

0     2018-12-14 12:41:33
1     2018-12-14 12:42:09
2     2018-12-14 12:47:36
3     2018-12-14 12:50:15
4     2018-12-14 12:50:31
              ...        
760   2020-03-06 20:12:51
761   2020-07-28 04:03:13
762   2020-07-28 04:03:20
763   2020-08-26 09:06:44
764   2020-09-21 01:47:10
Name: Timestamp, Length: 765, dtype: datetime64[ns]

The code above changes the timestamp datatype of the initial dataframe into a datetime. Does this affect new dataframes generated from the initial dataframe? Only one way to find out.

In [78]:
df_subset_1.dtypes

Timestamp                 object
Age                      float64
Gender                    object
City                      object
Position                  object
Years of experience      float64
Your level                object
Current Salary           float64
Main language at work     object
Company size              object
Company type              object
dtype: object

Apparently, no it doesn't. We therefore have to change it ourselves.

In [79]:
pd.to_datetime(df_subset_1["Timestamp"])

0     2018-12-14 12:41:33
1     2018-12-14 12:42:09
2     2018-12-14 12:47:36
3     2018-12-14 12:50:15
4     2018-12-14 12:50:31
              ...        
760   2020-03-06 20:12:51
761   2020-07-28 04:03:13
762   2020-07-28 04:03:20
763   2020-08-26 09:06:44
764   2020-09-21 01:47:10
Name: Timestamp, Length: 765, dtype: datetime64[ns]

During data preparation we may need to **rename the columns**. This might be the case when our column names have special characters, spaces or are in the wrong case. In our data, we'll want to remove the spaces in the column names and replace them with underscores.

In [80]:
for column in df_2018.columns:
    if " " in column:
        df_2018 = df_2018.rename(columns = {column: column.replace(" ", "_")})

df_2018.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', 'Are_you_getting_any_Stock_Options?',
       'Main_language_at_work', 'Company_size', 'Company_type'],
      dtype='object')

This could also be done by renaming each column, one by one using a dictionary, but I wanted to be fancy.

The non-fancy way:
```python
df_2018 = df_2018.rename(columns = 
    {"Years of experience": "Years_of_experience"
    "Your level": "Your_level"
    # and so on
    })
```

Note: You should have renamed this before you subsetted the data because, the columns of your subset still need to be renamed. See:

In [81]:
df_subset_1.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Main language at work', 'Company size',
       'Company type'],
      dtype='object')

Now you'll have to subset the data all over again

In [82]:
df_subset_1 = df_2018[
    ['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', #'Are_you_getting_any_Stock_Options?',
       'Main_language_at_work', 'Company_size', 'Company_type']
]

df_subset_1.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years_of_experience',
       'Your_level', 'Current_Salary', 'Salary_one_year_ago',
       'Salary_two_years_ago', 'Main_language_at_work', 'Company_size',
       'Company_type'],
      dtype='object')